poetry/console/commands/shell.py

```diff
@@ -0,0 +1,51 @@
+from os import environ
+
+from .venv_command import VenvCommand
+
+
+class ShellCommand(VenvCommand):
+    """
+    Spawns a shell within the virtual environment.
+
+    shell [options]
+    """
+
+    help = """The <info>shell</> command spawns a shell, according to the
+<comment>$SHELL</> environment variable, within the virtual environment.
+If one doesn't exist yet, it will be created.
+"""
+
+    def handle(self):
+        # This variable will be set if the current shell was spawned by poetry
+        current_venv = environ.get("POETRY_ACTIVE")
+
+        # Alternatively, a virtual environment might have already been
+        # activated in some other way
+        if not current_venv and not self.venv.is_venv():
+            current_venv = environ.get("VIRTUAL_ENV")
+
+        if current_venv:
+            self.line(
+                "Virtual environment already activated: "
+                "<info>{}</>".format(current_venv)
+            )
+            return
+
+        if self.venv._windows:
+            # Windows is probably more complicated than this
+            # and support to cmd alternatives should probably be added
+            shell = "cmd"
+            args = ["/k"]
+        else:
+            shell = environ.get("SHELL")
+            args = []
+            if shell is None:
+                self.error("The SHELL environment variable must be set.")
+                return
+
+        self.line("Spawning shell within <info>{}</>".format(self.venv.venv))
+
+        # Setting this to avoid spawning unnecessary nested shells
+        environ["POETRY_ACTIVE"] = str(self.venv.venv)
+        self.venv.execute(shell, *args)
+        environ.pop("POETRY_ACTIVE")
```

Comment:
I also don't get this piece of code... Why are you poping the variable? I thought we want it to stay since this function checks for it at the beggining, with this code, there is no way that on line 20 you'll find the env variable since it is not being set anywhere else then in this function at the end of which it is being removed from the end...

Reply:
We will be blocked at the `execute` line until the new shell is closed. Only then do we unset the variable.

Reply:
Ok, thanks for explaining. Do you plan to add some tests for this command as well?

Reply:
I think this feature is not very easily testable, but if you guys deem it necessary, I can give it a try.